### PR TITLE
Gated infrastructure deploy template

### PR DESCRIPTION
**Pull request by srjames-ukho**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,12 +4,13 @@ This repository contains template files that can be used in Azure DevOps Pipelin
...
+| [Github Tag](./github-tag)                                   | Github tags to your commits for marking significant points in a project's development                                                              |
+| [Retain pipeline run](./retain-pipelinerun)                  | Retaining a pipeline run for longer than the default configured 30 days after production release                                                   |
+| [SAST](./sast)                                               | SAST scan of the dotnet projects                                                                                                                   |
+| [tfsec](./tfsec)                                             | Static analysis security scanner against your Terraform                                                                                            |
+| [Gated Infrastructure Deploy](./gated-infrastructure-deploy) | Will trigger manual validation if terraform plan detects destroyed resources, if approved or no resources are being destroyed will terraform apply |
```
Comment (elerivaliant) [inline README.md:16]:
```suggestion
| [Gated Infrastructure Deploy](./gated-infrastructure-deploy) | Will trigger manual validation step if Terraform plan detects resource destruction |
```

File: gated-infrastructure-deploy/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
+# Template for Gated Infrastructure Deploy
+
+This template is designed to allow you to deploy terraform resources with a manually verification if any resources are getting destroyed. This manual verification will only trigger if there is indication that resources will be destroyed. If triggered, then the job will fail after a timeout has been reached. If there is no indication that resources wil be destroyed, the template will carry on and deploy the resources.
```
Comment (elerivaliant) [inline gated-infrastructure-deploy/README.md:3]:
```suggestion
This template is designed to allow you to deploy resources via Terraform with a manual verification step if the Terraform `plan` indicates that any resources will be destroyed. When no resources are flagged for destruction then the pipeline will automatically continue to Terraform `apply`. If the verification job is triggered, then you will need to approve in Azure DevOps in order to move on to the `apply` step. The job will fail if you either reject or do not approve within the timeout period.
```

File: gated-infrastructure-deploy/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    resources:
+      repositories:
+        - repository: UKHOTemplates
+          type: github
+          endpoint: UKHO
+          name: UKHO/devops-pipelinetemplates
+          ref: refs/heads/gated-infrastructure-deploy-template
+    ```
+
+2. Add this template block to your pipeline and fill out the values that relate to you deployment. The template content is a series of `jobs`, therefore they need to be rooted under a `stage`.
```
Comment (elerivaliant) [inline gated-infrastructure-deploy/README.md:23]:
```suggestion
2. Add this template block to your pipeline and fill out the values that relate to your deployment. The template content is a series of `jobs`, therefore they need to be rooted under a `stage`.
```

File: gated-infrastructure-deploy/README.md [hunk 4]
```diff
@@ -43,26 +55,88 @@ This template makes use of `terraform plan` and `terraform apply`.
...
+    - *TerraformWorkspace*: This value will be appended onto the blob name in the form `[TFStateBlobName]:[TerraformWorkspace]`, e.g. `terraform.deployment.tfplan:dev`   
 
-- the plan job
-- the manual verification
-- the apply job
+    - *TerraformArtifact*: The files in the artifact will be used without modifying their contents. Your pipeline will need to include jobs to make any required modfications to your supporting files e.g. inserting configuration values, before publishing the artifact.
+
+    - *TerraformArtifactConfigRelativePath*: Inside the template, the full path to the configuration will be `$(Pipeline.Workspace)/[TerraformArtifact][TerraformArtifactConfigRelativePath]`.
+      
+      Example 1: The artifact is named 'tfartifact' and the .tf files are in the root. `TerraformArtifactConfigRelativePath` would be `/` and the concatenated path would be `$(Pipeline.Workspace)/terraformartifact/`.
```
Comment (srjames-ukho) [inline gated-infrastructure-deploy/README.md:80]:
```suggestion
      Example 1: The artifact is named 'tfartifact' and the .tf files are in the root. `TerraformArtifactConfigRelativePath` would be `/` and the concatenated path would be `$(Pipeline.Workspace)/tfartifact/`.
```

Adding example name into the concatenated path.

File: gated-infrastructure-deploy/README.md [hunk 5]
```diff
@@ -43,26 +55,88 @@ This template makes use of `terraform plan` and `terraform apply`.
...
+
+      ```
+      +-- terraform
+      |   +-- azure.tf
+      |   +-- main.tf
+      |   +-- variables.tf
+      |   +-- output.tf
+      ```
```
Comment (srjames-ukho) [inline gated-infrastructure-deploy/README.md:88]:
```suggestion
      ```
      +-- tfartifact
      |   +-- azure.tf
      |   +-- main.tf
      |   +-- variables.tf
      |   +-- output.tf
      ```
```

Aligning with the example.

Comment (srjames-ukho) [inline gated-infrastructure-deploy/README.md:88]:
```suggestion
      ```
      +-- tfartifact
      |   +-- azure.tf
      |   +-- main.tf
      |   +-- variables.tf
      |   +-- output.tf
      ```
```

Aligning with the example.

File: gated-infrastructure-deploy/README.md [hunk 6]
```diff
@@ -43,26 +55,88 @@ This template makes use of `terraform plan` and `terraform apply`.
...
+      Example 2: The artifact is named 'buildartifact' and the .tf files are in a subfolder. `TerraformArtifactConfigRelativePath` would be `terraform/` and the concatenated path would be `$(Pipeline.Workspace)/buildartifact/terraform/`.
+
+      ```
+      +-- src
+      +-- terraform
+      |   +-- azure.tf
+      |   +-- main.tf
+      |   +-- variables.tf
+      |   +-- output.tf
+      ```
```
Comment (srjames-ukho) [inline gated-infrastructure-deploy/README.md:99]:
```suggestion
      ```
      +-- buildartifact
      |   +-- terraform
      |       +-- azure.tf
      |       +-- main.tf
      |       +-- variables.tf
      |       +-- output.tf
      ```
```
Updating to be inline with the example.

File: gated-infrastructure-deploy/README.md (hunk 5 above)
Comment (elerivaliant) [reply gated-infrastructure-deploy/README.md]: Difference between folder structure and artifact unclear, resolved by commit 45fe820

File: gated-infrastructure-deploy/README.md (hunk 6 above)
Comment (elerivaliant) [reply gated-infrastructure-deploy/README.md]: Difference between folder structure and artifact unclear, resolved by commit 45fe820

**Review by elerivaliant — APPROVED** [review]

**Pull request merged**